### DISCUSSION TITLE
ENOMEM on request('post'...) with 1.21.0 version

### DISCUSSION THREAD

**Discussion by abegines** [micropython/micropython#12788] [ESP32] [answered]
Hello,
I've a device reporting data through https to server (it's mandatory use https)

The the device is based on ESP32 devkit c1 v4, Chip is ESP32-D0WDQ6 (revision v1.0)

When i was on micropython `1.19.1` all works fine, with `1.20.0` also fine...
But if i upgrade to `1.21.0` i get ENOMEM exceptions inside wrap_socket when I call `urequests.request('post', ...`

For comparing memory usage and so, i write this function:

```
def meminfo(title):
    print(title, '===================================================')
    print(os.uname())
    print('mem_free:', gc.mem_free())
    print('mem_alloc:', gc.mem_alloc())
    micropython.mem_info()
    print(title, '------------------------------------------------------------------------')

```
Then I call it JUST AFTER connecting to the wifi and do gc.collect() and JUST BEFORE calling the very first request('post') with this block of code:

```
#wifi connected just now
gc.collect()
meminfo('First post after wifi connected')
response= urequests.request( 'POST', 'https://miservername.com/api/getEntorno', json = {"a":25})
print (response.status_code, response.text)
```

As said...

When I run on 1.19.1 it shows...

```
WIFI connected, trying to post... ==================================================================================================
(sysname='esp32', nodename='esp32', release='1.19.1', version='v1.19.1 on 2022-06-18', machine='ESP32 module with ESP32')
mem_free: 50096
mem_alloc: 61072
stack: 1264 out of 15360
GC: total: 111168, used: 61072, free: 50096
 No. of 1-blocks: 596, 2-blocks: 175, max blk sz: 282, max free sz: 361
WIFI connected, trying to post... --------------------------------------------------------------------------------------------------
200 {"err":false,"result":{"entorno":"PROD","horaz":"2023-10-24T08:54:19.494Z","horaoffset":-120}}
```

When i run on 1.20.0 it shows...

```
WIFI connected, trying to post... ===========================================
(sysname='esp32', nodename='esp32', release='1.20.0', version='v1.20.0 on 2023-04-26', machine='ESP32 module with ESP32')
mem_free: 49728
mem_alloc: 61440
stack: 1264 out of 15360
GC: total: 111168, used: 61440, free: 49728
 No. of 1-blocks: 609, 2-blocks: 174, max blk sz: 282, max free sz: 365
WIFI connected, trying to post... ----------------------------------------------------------
200 {"err":false,"result":{"entorno":"PROD","horaz":"2023-10-24T08:57:12.151Z","horaoffset":-120}}
```

When I run on 1.21.0 it shows...

```
WIFI connected, trying to post... ============================================
(sysname='esp32', nodename='esp32', release='1.21.0', version='v1.21.0 on 2023-10-05', machine='Generic ESP32 module with ESP32')
mem_free: 85920
mem_alloc: 60512
stack: 1264 out of 15360
GC: total: 128000, used: 60512, free: 67488, max new split: 18432
 No. of 1-blocks: 577, 2-blocks: 173, max blk sz: 282, max free sz: 2638
WIFI connected, trying to post... -------------------------------------------------------------------------
[Errno 12] ENOMEM
```

Suprisingly on 1.21.0 we have 128000 of GC total, instead of 111168 in previous versions... but it fails when trying to post with https.
Have you any cue about that?

Thanks!

**Comment by jimmo** [comment]
See also https://github.com/orgs/micropython/discussions/12791 (@bobveringa) which I think might be related.

**Comment by jimmo** [comment, marked as answer]
I think what's going on here is a variation on https://github.com/micropython/micropython/issues/8940 and we hoped would be improved in v1.21 by the changes described in https://github.com/orgs/micropython/discussions/12316 

The high-level issue is that on ESP32, the RAM is available to us and TLS (mbedtls) as disjoint regions of contiguous RAM. The old behavior was that MicroPython would allocate the largest such region and leave everything else for the IDF (wifi, ssl, socket buffers, etc). Different IDF versions resulted in a different memory layout as their internal allocations and data structures moved around (this is one major reason for differences in v1.21).

The problem we have now is that MicroPython can actually allocate more RAM than it could before (as with the split heap feature, it can grow past the first contiguous region into the next one). Which means that there might be not enough left in the IDF heap for ssl buffers.

The only real solution to this is to pre-allocate the SSL buffers, so you can allocate them at startup. This is something we'd like to support in v1.22, and was a big part of the motivation for introducing the SSLContext class.

For a workaround... I'm not entirely sure. I think we'd need to get to bottom of why the IDF has no more ram available (i.e. is it because MicroPython has used up all available IDF heap into multiple splits). Are you able to get your program to call `micropython.mem_info()` before each call to wrap_socket (or as close as possible to the call to wrap_socket). That will confirm this.

The best thing I can suggest is to either explicitly call `gc.collect()` more frequently, during program startup and execution, which will put pressure on the heap to grow less quickly and fragment less. Equivalently, setting `gc.threshold` will automatically collect. That said, it wont allocate a new split without first doing a gc.collect().

But really what we need is the pre-allocated SSL buffers.

@projectgus Might have more thoughts on this.

**Reply by abegines** [reply]
Hi, 
`micropython.mem_info()` called just before firrst call to `warp_socket`. With and without previous `gc.collect()`
but both throw ENOMEM...

`micropython.mem_info()` JUST before `wrap_socket`
```
stack: 1952 out of 15360
GC: total: 128000, used: 60416, free: 67584, max new split: 17408
 No. of 1-blocks: 608, 2-blocks: 175, max blk sz: 282, max free sz: 2793
[Errno 12] ENOMEM
```


`micropython.mem_info() `JUST before `wrap_socket`, with previous `gc.collect()`
```
stack: 1952 out of 15360
GC: total: 128000, used: 60144, free: 67856, max new split: 17408
 No. of 1-blocks: 590, 2-blocks: 174, max blk sz: 282, max free sz: 2893
[Errno 12] ENOMEM
```

**Reply by dpgeorge** [reply]
It looks like your application needs more than 64k of heap, and so in v1.21.0 the heap must expand to two separate pools, the initial 64k pool and then a second one.  This gives your application a MicroPython heap of 128000 bytes, which is about 10000 bytes more than in v1.20.0.

Unfortunately it seems that this 10000 extra MicroPython heap is taking just a bit too much from the IDF and not leaving enough for an SSL connection.

@jimmo maybe this is a common enough case (esp32-og, application needs just a little more than 64k MicroPython heap, and also wants to make one SSL connection) that we should tune it so that the second pool that's added to the heap leaves enough IDF heap for one SSL buffer?  Or should we just wait for pre-allocated SSL buffers?

**Reply by projectgus** [reply]
Quick thought, but we could ease off the "greedy" doubling as we start to approach the memory limit, as well.

**Reply by dpgeorge** [reply]
Should be fixed now by f6d630877c3b31a5091dff2fd615ddb92189a46c: the heap is tuned so expanding it once still allows to connect to WiFi and make an SSL request.

**Comment by bobveringa** [comment]
In my application, heap fragmentation was a big issue before, so the garbage collection is already on the heavier side. During startup (when many resources are created and cleaned up) the garbage collector threshold is set at 1024, which slows down the system a bit, but significantly reduces heap fragmentation. 

For my use-case, I ended up finding a user c module that had a fairly large lookup table (which was nice for performance). Deleting these tables resolved my issues with SSL connections. These tables were very nice to have, so this is not ideal. I'll continue my investigation into what is happening and if any of my c user modules that allocate dynamic memory are also impacted.

This is the output of `micropython.mem_info()` before the socket is wrapped:
```
-- with lookup tables ~8k
stack: 2848 out of 15360
GC: total: 128000, used: 82704, free: 45296, max new split: 43008
 No. of 1-blocks: 1918, 2-blocks: 216, max blk sz: 105, max free sz: 2447

-- without lookup tables
stack: 2848 out of 15360
GC: total: 128000, used: 88080, free: 39920, max new split: 47104
 No. of 1-blocks: 2257, 2-blocks: 218, max blk sz: 105, max free sz: 2464
```

**Comment by projectgus** [comment]
I've added a tweak here to try not to prematurely exhaust the free system heap:
https://github.com/micropython/micropython/pull/13035

@abegines @bobveringa Are you able to try this version with your applications and see if it helps?

**Reply by bobveringa** [reply]
Certainly looks promising, I'll try to figure out how I can run the version on my devices. I'm a bit time constraint right now, but I'll try to make some time somewhere next week.

**Comment by smudge1977** [comment]
I have just been chasing an ENOMEM issue is my app...
`sysname='rp2', nodename='rp2', release='1.21.0', version='v1.21.0 on 2023-10-06 (GNU 13.2.0 MinSizeRel)', machine='Raspberry Pi Pico W with RP2040'`

This got me on the right path: https://forum.micropython.org/viewtopic.php?t=3499

So in my app I called `free()` in my main loop...  Now interestingly even this you see your free memory slowly dropping!!

In my app I am using `urequest` and the response has to go into memory to be processed - I have an object for a card payment provider and there are some json responses that move around as you would expect but it seems the garbage collector is not cleaning up!

So on my PI Pico W the main loop runs at ~54% free...
```
INFO:__main__:Loop...
Disk: 0.5039063 MB Memory: Total:177280 Free:38064 (21.47%)
fINFO:__main__:Loop...
Disk: 0.5039063 MB Memory: Total:177280 Free:36880 (20.80%)
INFO:__main__:Waiting on payment
INFO:__main__:Loop...
Disk: 0.5039063 MB Memory: Total:177280 Free:15264 (8.61%)
INFO:__main__:Waiting on payment
DEBUG:__main__:retrieve payment pi_3OddcPLRpaTJzNCb1MtqyoAO
CRITICAL: retrieve_payment_intent  [Errno 12] ENOMEM
CRITICAL:__main__:Main loop finally - bye
```

```py
import gc
import os
import time


def df():
    s = os.statvfs('//')
    return ('{0} MB'.format((s[0]*s[3])/1048576))


def free(full=False, collect=False):
    if collect:
        gc.collect()
    F = gc.mem_free()
    A = gc.mem_alloc()
    T = F+A
    P = '{0:.2f}%'.format(F/T*100)
    if not full:
        return P
    else:
        return ('Total:{0} Free:{1} ({2})'.format(T, F, P))

while True:
    print(free(True, False))
    time.sleep(1)

```


Now the odd thing is if I use the above code and print `free(True, True)` to do a garbabge collect then everything is all fine.

**Reply by projectgus** [reply]
> Now the odd thing is if I use the above code and print free(True, True) to do a garbabge collect then everything is all fine.

Something is probably fragmenting the Python heap, to the point where a large allocation is failing with ENOMEM. Running garbage collection more frequently prevents it from fragmenting, so there's still enough contiguous space for the large allocation when needed.

If you run `micropython.mem_info()` then this should give confirmation as to whether the problem is fragmentation (the `GC: free:` number may stay high but `max free sz:` number will trend down - that second number is blocks, not bytes).

If you're not able to restructure the application to avoid a regular large allocation, the most common way around this is to set [gc.threshold()](https://docs.micropython.org/en/latest/library/gc.html#gc.threshold) and find a value where the GC runs often enough that fragmentation no longer gets out of hand.
